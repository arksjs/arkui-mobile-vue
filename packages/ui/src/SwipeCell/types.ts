import type { StateType } from '../helpers'

export type ButtonOption = {
  text: string
  type?: StateType
}

export type OnButtonClick = (payload: {
  item: Required<ButtonOption>
  index: number
}) => void

export interface SwipeCellProps {
  buttons: ButtonOption[]
}

export interface SwipeCellEmits {
  onButtonClick?: OnButtonClick
}

export type {
  ButtonOption as SwipeCellButtonOption,
  OnButtonClick as SwipeCellOnButtonClick
}
