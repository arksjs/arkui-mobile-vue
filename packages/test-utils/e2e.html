<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
