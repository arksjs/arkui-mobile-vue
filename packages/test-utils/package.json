{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
