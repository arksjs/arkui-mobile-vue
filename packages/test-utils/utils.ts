export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
