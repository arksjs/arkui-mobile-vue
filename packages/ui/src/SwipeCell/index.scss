@import '../style/var.scss';
@import '../style/mixin.scss';

.ak-swipe-cell {
  --ak-color: #{$primary-color};
  --ak-dark-color: var(--ak-color-primary-dark);

  cursor: grab;
  overflow: hidden;

  + .ak-swipe-cell::before {
    content: '';
    margin-left: 16px;
  }

  &_inner {
    transition: transform 0s ease;
    position: relative;
  }

  &_buttons {
    position: absolute;
    top: 0;
    height: 100%;
    white-space: nowrap;
    right: 0;
    transform: translate3d(100%, 0, 0);
  }

  &_button {
    display: inline-flex;
    vertical-align: top;
    align-items: center;
    padding: 0 12px;
    position: relative;
    font-size: 16px;
    height: 100%;
    margin: 0;
    border: none;
    outline: 0;
    border-radius: 0;
    cursor: pointer;
    background-color: var(--ak-color);
    color: #fff;
    white-space: pre-line;
    overflow: visible;
    transition: transform 0s ease;

    &:active {
      background-color: var(--ak-dark-color);
    }

    &::before {
      content: '';
      position: absolute;
      right: 0;
      top: 0;
      height: 100%;
      width: 100vw;
      transform: translate3d(100%, 0, 0);
      background-color: var(--ak-color);
    }

    &.type--success,
    &.type--success::before {
      --ak-color: #{$success-color};
      --ak-dark-color: var(--ak-color-success-dark);
    }

    &.type--warning,
    &.type--warning::before {
      --ak-color: #{$warning-color};
      --ak-dark-color: var(--ak-color-warning-dark);
    }

    &.type--danger,
    &.type--danger::before {
      --ak-color: #{$danger-color};
      --ak-dark-color: var(--ak-color-danger-dark);
    }
  }
}
