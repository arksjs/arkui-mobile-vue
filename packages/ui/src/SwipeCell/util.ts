import { STATE_TYPES, getEnumsValue, isString } from '../helpers'
import type { ButtonOption } from './types'

export const getButtons = (buttons?: ButtonOption[]) => {
  const _buttons: Required<ButtonOption>[] = []

  if (Array.isArray(buttons)) {
    buttons.forEach(v => {
      if (v && isString(v.text)) {
        _buttons.push({
          text: v.text,
          type: getEnumsValue(STATE_TYPES, v.type)
        })
      }
    })
  }

  return _buttons
}

export const getInnerStyles = ({
  translateX,
  duration
}: {
  translateX: number
  duration: number
}) => ({
  transform: 'translate3d(-' + translateX + 'px, 0px, 0px)',
  transitionDuration: duration + 's'
})

export const getButtonStyles = ({
  buttonTranslateXs,
  duration,
  index
}: {
  buttonTranslateXs: number[]
  duration: number
  index: number
}) => ({
  transform:
    'translate3d(-' + (buttonTranslateXs[index] || 0) + 'px, 0px, 0px)',
  transitionDuration: duration + 's'
})
