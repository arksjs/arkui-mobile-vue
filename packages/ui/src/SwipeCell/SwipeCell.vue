<template>
  <div class="fx-swipe-cell fx-horizontal-hairline" ref="root">
    <div class="fx-swipe-cell_inner" :style="innerStyles">
      <slot></slot>
      <div class="fx-swipe-cell_buttons" ref="buttonsEl">
        <button
          class="fx-swipe-cell_button"
          :class="['type--' + item.type]"
          v-for="(item, index) in buttons2"
          :key="index"
          :style="
            getButtonStyles({
              buttonTranslateXs,
              duration,
              index
            })
          "
          @click="onButtonClick(item, index)"
        >
          {{ item.text }}
        </button>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { ref, defineComponent, computed, reactive } from 'vue'
import type { PropType } from 'vue'
import { cloneData, rangeNumber, noop } from '../helpers/util'
import type { ButtonOption, SwipeCellEmits } from './types'
import { useTouch } from '../hooks/use-touch'
import { getStretchOffset } from '../helpers/animation'
import type { PropsToEmits } from '../helpers/types'
import { getButtons, getInnerStyles, getButtonStyles } from './util'
import { useStop, useBlur } from '../hooks/use-event'

interface SwipeCellCoords {
  startX: number
  buttonsW: number
  isShow: boolean
  isSwipe: boolean
}

export default defineComponent({
  name: 'fx-swipe-cell',
  props: {
    buttons: {
      type: Array as PropType<ButtonOption[]>,
      validator: (items: ButtonOption[]) => {
        return (
          Array.isArray(items) &&
          items.filter(item => {
            return !(item && typeof item.text === 'string')
          }).length === 0
        )
      },
      default: () => [] as ButtonOption[]
    }
  },
  emits: {
    buttonClick: payload => payload && typeof payload.index === 'number'
  } as PropsToEmits<SwipeCellEmits>,
  setup(props, ctx) {
    const root = ref<HTMLElement>()
    const buttonsEl = ref<HTMLElement>()
    const translateX = ref(0)
    const duration = ref(0)
    const buttonTranslateXs = reactive<number[]>([])
    let coords: SwipeCellCoords | null
    let isShow = false

    function show(x: number) {
      translateX.value = x
      duration.value = 0.6

      buttonTranslateXs.forEach((v, k) => {
        buttonTranslateXs[k] = 0
      })

      isShow = true
    }

    function hide() {
      if (!isShow) {
        return
      }

      isShow = false

      translateX.value = 0
      duration.value = 0.6

      buttonTranslateXs.forEach((v, k) => {
        buttonTranslateXs[k] = 0
      })
    }

    function onButtonClick(item: Required<ButtonOption>, index: number) {
      ctx.emit('buttonClick', {
        item: cloneData(item),
        index
      })
      hide()
    }

    const buttons2 = computed(() => getButtons(props.buttons))

    useTouch({
      el: root,
      onTouchStart(e) {
        if (props.buttons.length === 0) {
          return
        }

        coords = {
          startX: e.touchObject.clientX,
          buttonsW: (buttonsEl.value as HTMLElement).clientWidth,
          isShow: translateX.value > 0,
          isSwipe: false
        }

        if (coords.isShow) {
          e.stopPropagation()
        }
      },
      onTouchMove(e) {
        if (!coords) {
          return
        }

        const { startX, buttonsW, isSwipe, isShow } = coords

        let x = startX - e.touchObject.clientX

        if (!isShow && !isSwipe && x < 0) {
          coords = null
          return
        }
        coords.isSwipe = true

        if (isShow) {
          x += buttonsW
        }

        const max = rangeNumber(x, 0, buttonsW)

        const $children = (buttonsEl.value as HTMLElement).children

        for (let i = 0, len = $children.length; i < len; i++) {
          buttonTranslateXs[i] =
            (($children[i] as HTMLElement).offsetLeft * (buttonsW - max)) /
            buttonsW
        }

        translateX.value =
          max + (x > buttonsW ? getStretchOffset(x - buttonsW) : 0)
        duration.value = 0

        e.stopPropagation()
      },
      onTouchEnd(e) {
        if (coords) {
          const { isSwipe, buttonsW } = coords

          if (isSwipe && translateX.value > buttonsW / 2) {
            // 展示
            show(buttonsW)
          } else {
            hide()
          }

          coords = null
          e.stopPropagation()
        }
      }
    })

    const innerStyles = computed(() =>
      getInnerStyles({
        translateX: translateX.value,
        duration: duration.value
      })
    )

    useBlur(hide)
    useStop(buttonsEl, 'click')

    return {
      root,
      buttonsEl,
      buttonTranslateXs,
      buttons2,
      duration,
      noop,
      onButtonClick,
      innerStyles,
      getButtonStyles
    }
  }
})
</script>
