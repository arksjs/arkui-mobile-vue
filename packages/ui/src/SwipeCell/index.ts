import { withInstall } from '../helpers/with-install'
import SwipeCell from './SwipeCell.vue'

export { SwipeCell }
export const install = withInstall(SwipeCell)
export default SwipeCell
