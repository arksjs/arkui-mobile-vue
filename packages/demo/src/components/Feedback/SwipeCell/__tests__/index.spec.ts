import { shallowMount } from '@vue/test-utils'
import { SwipeCell } from '@/SwipeCell'

describe('SwipeCell', () => {
  test('should render default slot correctly', () => {
    const wrapper = shallowMount(SwipeCell, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render set buttons correctly', () => {
    const wrapper = shallowMount(SwipeCell, {
      props: {
        buttons: [
          {
            text: '加入收藏',
            type: 'warning'
          },
          {
            text: '删除',
            type: 'danger'
          }
        ]
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
