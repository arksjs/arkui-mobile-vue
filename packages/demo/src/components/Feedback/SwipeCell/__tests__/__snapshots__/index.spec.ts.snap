// Vitest Snapshot v1

exports[`SwipeCell > should render default slot correctly 1`] = `
"<div class=\\"fx-swipe-cell fx-horizontal-hairline\\">
  <div class=\\"fx-swipe-cell_inner\\" style=\\"transform: translate3d(-0px, 0px, 0px); transition-duration: 0s;\\">
    <div class=\\"fx-swipe-cell_buttons\\"></div>
  </div>
</div>"
`;

exports[`SwipeCell > should render set buttons correctly 1`] = `
"<div class=\\"fx-swipe-cell fx-horizontal-hairline\\">
  <div class=\\"fx-swipe-cell_inner\\" style=\\"transform: translate3d(-0px, 0px, 0px); transition-duration: 0s;\\">
    <div class=\\"fx-swipe-cell_buttons\\"><button class=\\"fx-swipe-cell_button type--warning\\" style=\\"transform: translate3d(-0px, 0px, 0px); transition-duration: 0s;\\">加入收藏</button><button class=\\"fx-swipe-cell_button type--danger\\" style=\\"transform: translate3d(-0px, 0px, 0px); transition-duration: 0s;\\">删除</button></div>
  </div>
</div>"
`;
